### ISSUE TITLE
Source generators can't access binary files

### ISSUE THREAD

**Issue by CptWesley** [dotnet/roslyn#47292]
As far as I can see, source generators only have access to files specified as `AdditionalFiles`. However, this means that binary files can't be used for generating code. Adding a binary file as an `AdditionalFiles` will result in a `CS2015` error. I've personally had some use cases for generating source code based on binary files, so I'm curious if this was intended, or alternatively, if I missed something in the specifications.

**Comment by RikkiGibson** [comment]
@chsienki @jasonmalinowski this feels slightly odd, should the compiler care whether or not a user's AdditionalFile is in a binary format or not?

**Comment by CyrusNajmabadi** [comment]
> should the compiler care whether or not a user's AdditionalFile is in a binary format or not?

I would expect so.  Don't we expose this through Document/SourceText.  These are *string* oriented abstractions.  That's why they need encodings and whatnot.  We'd need an entirely new binary abstraction for roslyn (through the entire stack) to support this.

**Comment by jasonmalinowski** [comment]
Indeed, we'd need a non-SourceText abstraction. If we're restricting it to just additional files that won't be too bad since there's already an AdditionalText file that we handed around that lets you get the SourceText: 

https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/DiagnosticAnalyzer/AdditionalText.cs

So we could easily add some sort of API to that to get the binary blob. You just have to plug your nose adding such an API to a type called AdditionalText...

**Comment by chsienki** [comment]
Related: https://github.com/dotnet/roslyn/pull/59578/
